### ISSUE TITLE
How to marshal integer as boolean

### ISSUE THREAD

**Issue by Alovchin91** [mono/CppSharp#1205]
There is little documentation on the matter, and I couldn't find a way to make it generate such a code.

Imagine that I have a typedef like:

```cpp
typedef int boolean_t;
```

And then some functions:

```cpp
void set_bool_value(boolean_t value);
boolean_t get_bool_value();
```

By default, CppSharp would generate prototypes like:

```cs
void SetBoolValue(int value);
int GetBoolValue();
```

I would like to generate the following code instead:

```cs
void SetBoolValue([MarshalAs(UnmanagedType.I4)] bool value);
[return: MarshalAs(UnmanagedType.I4)] bool GetBoolValue();
```

I tried using TypeMaps to achieve this, but CppSharp doesn't seem to call into the implemented TypeMap at all.

Could you please help me to find an appropriate solution? Thanks!

**Comment by tritao** [comment]
A type map would indeed be the most appropriate solution here.

We check for the existence of a type map for typedefs here: https://github.com/mono/CppSharp/blob/master/src/Generator/Generators/CSharp/CSharpTypePrinter.cs#L265

Maybe you can add a breakpoint here and see why its not found for `boolean_t`.

Alternatively, you can create a pass that checks every function declaration, and if the return type or a parameter type is `boolean_t` then replaces it with a bool type.

However its not as elegant, and you have to be careful to replace it with an equivalently sized type, or the ABI might have a different size.

**Comment by Alovchin91** [comment]
Hi @tritao, thanks for the response!

I tried to manually call that method `Context.TypeMaps.FindTypeMap` for the `TypedefType.Declaration.Type`, and what it finds is the TypeMap of an integer type (apparently the first one that exists in the database).

My TypeMap is defined as follows:

```cs
[TypeMap("boolean_t")]
public class BooleanTypeMap : TypeMap
{
    ...
}
```

There is zero documentation on how to use type maps, so I found this solution somewhere in tests.

Also, I've tried to create a pass to check for `boolean_t` and replace it with a bool. but I have no clue on how to correctly marshal it (that is, add a `MarshalAs(UnmanagedType.I4)` attribute).

It seems for me now that there is no way to work around this fail.

**Comment by Alovchin91** [comment]
Any suggestions on how to achieve this? Maybe there is a way to access code writer from the TranslationUnitPass to manually write the attribute?

**Comment by tritao** [comment]
Yes, there is a way to access the code writer but not from the pass. Only via a type map. 

Can you try to debug why the type map is not found? Compile the source code and add a breakpoint on the line I suggested and you should be able to figure out. If you get stuck, feel free to stop by Gitter and I will try to help.

**Comment by Alovchin91** [comment]
@tritao I called that method manually and now I re-checked on manually built CppSharp in that CSharpTypePrinter class.

The TypeMap returned from the TypeMapDatabase is `{CppSharp.Types.Std.Int}` and not the typemap I defined. This is because it looks for a typemap based on the `decl.Type`, and it resolves to built-in `int` type.

Here is the small test project to reproduce this behavior:

[CppSharpTypeMapBug.zip](https://github.com/mono/CppSharp/files/3133170/CppSharpTypeMapBug.zip)

Apparently adding this line to the condition solves the problem:

```cs
if (** TypeMapDatabase.FindTypeMap(decl.OriginalName, out typeMap) || ** <-- This line
    TypeMapDatabase.FindTypeMap(decl.Type, out typeMap))
{
    ...
}
```

But maybe it's me who's doing something wrong.

@tritao Could you please take a look? Thanks!

**Comment by tritao** [comment]
I think the issue is that we need to check with the typedef type before trying with the typedef inner type.

Can you try replacing your new line with `TypeMapDatabase.FindTypeMap(typedef, out typeMap)` instead?

**Comment by Alovchin91** [comment]
@tritao Tried that and it doesn't work.

In this case it call the following overload:

```cs
public bool FindTypeMap(Declaration declaration, out TypeMap typeMap) =>
    FindTypeMap(new TagType(declaration), out typeMap);
```

And the TagType created is for `int` type: `"int" ["TagType"]`

**Comment by tritao** [comment]
Took a look, I think the real issue is that `TypedefType` comparison is buggy and does not take the original name into account. Had some issues quickly getting it to work, need to check it again later.

**Comment by tritao** [comment]
The comparison was buggy, but the real issue was that we were looking for a typemap with a resolved typedef before trying the actual typedef.

**Comment by Alovchin91** [comment]
@tritao I tried the updated build, now the TypeMap works as expected. Thanks!

I also noticed another thing: if you have that type as a `ref` parameter, the TypeMap is only being asked for `CSharpSignatureType` and not for `CSharpMarshalToManaged` or `CSharpMarshalToNative`.

So, consider the following scenario:

```cpp
typedef int boolean_t;

void get_value(boolean_t *pValue, boolean_t defaultValue);
```

TypeMap is as follows:

```cs
[TypeMap("boolean_t", GeneratorKind = GeneratorKind.CSharp)]
public class BooleanTypeMap : TypeMap
{
    public override CppSharp.AST.Type CSharpSignatureType(TypePrinterContext ctx)
    {
        return new CILType(
            ctx.Kind == TypePrinterContextKind.Managed
                ? typeof(bool)
                : typeof(int));
    }

    public override void CSharpMarshalToManaged(CSharpMarshalContext ctx)
    {
        ctx.Return.Write("({0} != 0)", ctx.ReturnVarName);
    }

    public override void CSharpMarshalToNative(CSharpMarshalContext ctx)
    {
        ctx.Return.Write("({0} ? 1 : 0)", ctx.Parameter.Name);
    }
}
```

In this case, the generated code will look something like:

```cs
...
internal static extern void get_value(int* pValue, int defaultValue);
...
public static void GetValue(ref bool pValue, bool defaultValue)
{
    fixed (bool* __pValue1 = &pValue)
    {
        var __arg0 = __valuep1;
        var __arg1 = (defaultValue ? 1 : 0);
        var __ret = __Internal.get_value(__arg0, __arg1);
    }
}
```

It cannot be compiled of course because `bool*` is not compatible with `int*`. Unfortunately, for `arg0` the TypeMap is not being called, so there is no way to fix that parameter in the TypeMap.

The workaround is to use TranslationUnitPass of course, but is this a bug or a known limitation?

**Comment by tritao** [comment]
Probably a bug or just something we never considered.
I will discuss the issue later with @ddobrev and see about a solution.
